src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,887 @@
+  // NOTICE: we can separate into disjoint and exceed. But currently maybe it is enough.
+  double Disjoint(Genome& genome1, Genome& genome2) {
+    double numDisjoint = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      bool linkContainedInGenome2 = genome2.ContainLink(innovId);
+      if (!linkContainedInGenome2) {
+        ++numDisjoint;
+      } 
+    }
+
+    for (ssize_t i=0; i<genome2.NumLink(); ++i) {
+      ssize_t innovId = genome2.aLinkGenes[i].InnovationId();
+      bool linkContainedInGenome1 = genome1.ContainLink(innovId);
+      if (!linkContainedInGenome1) {
+        ++numDisjoint;
+      }
+    }
+
+    ssize_t largerGenomeSize = std::max(genome1.NumLink(), genome2.NumLink());
+    double deltaD = numDisjoint / largerGenomeSize;
+    return deltaD; 
+  }
+
+  // Measure two genomes' weight difference.
+  double WeightDiff(Genome& genome1, Genome& genome2) {
+    double deltaW = 0;
+    ssize_t coincident = 0;
+
+    for (ssize_t i=0; i<genome1.NumLink(); ++i) {
+      int linkEnabledInGenome1 = (int) genome1.aLinkGenes[i].Enabled();
+      ssize_t innovId = genome1.aLinkGenes[i].InnovationId();
+      ssize_t idx = genome2.GetLinkIndex(innovId);
+      bool linkContainedInGenome2 = (idx != -1);
+
+      if (linkContainedInGenome2) {
+        int linkEnabledInGenome2 = (int) genome2.aLinkGenes[idx].Enabled();
+        deltaW += std::abs(genome1.aLinkGenes[i].Weight() * linkEnabledInGenome1 - 
+                           genome2.aLinkGenes[idx].Weight() * linkEnabledInGenome2);
+        ++coincident;
+      }
+    }
+
+    deltaW = deltaW / coincident;
+    return deltaW;
+  }
+
+  // Whether two genome belong to same species or not.
+  bool IsSameSpecies(Genome& genome1, Genome& genome2) {
+    double deltaD = Disjoint(genome1, genome2);
+    double deltaW = WeightDiff(genome1, genome2);
+    double delta = aCoeffDisjoint * deltaD + aCoeffWeightDiff * deltaW;
+
+    if (delta < aCompatThreshold) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Add genome to existing species or create new species.
+  void AddGenomeToSpecies(Population& population, Genome& genome) {
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      if (population.aSpecies[i].aGenomes.size() > 0) {
+        if (IsSameSpecies(population.aSpecies[i].aGenomes[0], genome)) {  // each first genome in species is the representative genome.
+          population.aSpecies[i].AddGenome(genome);
+          return;
+        }
+      }
+    }
+
+    Species newSpecies = Species();
+    newSpecies.AddGenome(genome);
+    newSpecies.StaleAge(0);
+    population.AddSpecies(newSpecies);
+  }
+
+  // Remove stale species.
+  void RemoveStaleSpecies(Population& population) {
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      if (population.aSpecies[i].StaleAge() > aStaleAgeThreshold) {
+        population.RemoveSpecies(i);
```

Comment:
This function could cause an error, because we can't increase `i` if we remove a species:

```
  void RemoveStaleSpecies(Population& population)
  {
    for (std::vector<Species>::iterator it = population.aSpecies.begin();
         it != population.aSpecies.end();  /*it++*/)
    {
      if(it->StaleAge() > aStaleAgeThreshold)
      {
        it = population.aSpecies.erase(it);
      }
      else
      {
        ++it;
      }
    }
  }
```